src/main/java/org/opensearch/knn/index/codec/KNNCodecVersion.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.apache.lucene.backward_codecs.lucene91.Lucene91Codec;
+import org.apache.lucene.backward_codecs.lucene92.Lucene92Codec;
+import org.apache.lucene.codecs.Codec;
+import org.apache.lucene.codecs.lucene94.Lucene94Codec;
+import org.apache.lucene.codecs.perfield.PerFieldKnnVectorsFormat;
+import org.opensearch.index.mapper.MapperService;
+import org.opensearch.knn.index.codec.KNN80Codec.KNN80CompoundFormat;
+import org.opensearch.knn.index.codec.KNN80Codec.KNN80DocValuesFormat;
+import org.opensearch.knn.index.codec.KNN910Codec.KNN910Codec;
+import org.opensearch.knn.index.codec.KNN920Codec.KNN920Codec;
+import org.opensearch.knn.index.codec.KNN920Codec.KNN920PerFieldKnnVectorsFormat;
+import org.opensearch.knn.index.codec.KNN940Codec.KNN940Codec;
+import org.opensearch.knn.index.codec.KNN940Codec.KNN940PerFieldKnnVectorsFormat;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Abstraction for k-NN codec version, aggregates all details for specific version such as codec name, corresponding
+ * Lucene codec, formats including one for k-NN vector etc.
+ */
+@AllArgsConstructor
+@Getter
+public enum KNNCodecVersion {
```

Comment:
Not quite sure if this enum class is required.
For now, it seems like there is 1:1 mapping between each enum type and Codec class. Then, what would be the advantage of having this enum class vs defining values in each Codec class itself?

Reply:
Can we have one generic codec class and generate each codec based on what is defined in this enum?

Reply:
We can't use one class for every version due to limitations from SPI. As per contract service class must have public no args constructor, it's instantiated based on [registry](https://github.com/opensearch-project/k-NN/blob/main/src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec) using either Class.forName or class.newInstance(). 
I think what we can do is pulling common logic to some base class to increase code re-usage

Reply:
Got it. Then, can we just have these values in each codec class itself? I don't see much value of having this enum class.

Reply:
For me the value is simpler up-version process. [Recent up-version change](https://github.com/opensearch-project/k-NN/commit/285589c770108ffefc85725f21295254d8be3cc0) for 3.0.0 requires way to many change IMO. In this change I want to 1 - minimize classes with lucene specific imports, 2 - have single place that knows "current" codec version. With 3 different codec classes lucene dependencies will be in each class. Current version can be initialized directly, but this will not use polymorphism, and we need some extra mechanism for this, like a factory approach that is currently in the main branch

Reply:
1 - Checked the Recent up-version change for 3.0.0. I think this whole PR might simplify the code change required but not by the introduction of this enum.
2 - We can have single place to hold current code version regardless using enum or not.
KNNCodecVersion.V_9_4_0 is used only by KNN940Codec class, and KNNCodecVersion.V_9_2_0 by KNN920Codec class. I am still not sure if it is good to have a separate enum in this case.

Reply:
Any plan for introducing basic codec class to remove more code duplication as suggested in previous comment?

Reply:
yes, I'll do that with next commit